File: Java/Cristian Maldonado/Matemática/Number_Of_Digits_Factorial_Optimized.java
public class Number_Of_Digits_Factorial_Optimized {

    static double M_E = 2.71828182845904523536;
    static double M_PI = 3.141592654;
    
    static long findDigits(int n){
        if(n < 0){
            return 0;
        }
        if(n <= 1){
            return 1;
        }
        double x = (n * Math.log10(n / M_E)+ Math.log10(2 * M_PI * n) / 2.0);
        return (long) Math.floor(x) + 1;
    }
    
    public static void main(String[] args) {
        System.out.println(findDigits(1));
        System.out.println(findDigits(50000000));
        System.out.println(findDigits(1000000000));
        System.out.println(findDigits(120));
    }    
}

File: Codigos estudiantes por lenguaje/C++/Cristian Maldonado/Matemática/Number_Of_Digits_Factorial_Optimized.cpp
#include <iostream>
#include <math.h>
//----------//
#define Euler 2.71828182845904523536
#define PI 3.141592654

using namespace std;

typedef long long int lli;

lli findDigits(lli n){
    if(n < 0){
        return 0;
    }
    if(n <= 1){
        return 1;
    }
    double x = (n * log10(n / Euler) + log10(2 * PI * n) / 2.0);
    return (lli)(floor(x) + 1);
}

int main()
{
    ios_base::sync_with_stdio(false); cout.tie(NULL);
    cout<<findDigits(1)<<endl;
    cout<<findDigits(50000000)<<endl;
    cout<<findDigits(1000000000)<<endl;
    cout<<findDigits(120)<<endl;
    return 0;
}
